README.md
# 一般说明

基础地址（`base-api`）为
* `https://hrznc.haier.net` 或
* `http://120.24.38.2:8080`

对请求的通用要求：
* 所有请求（包括对静态资源的请求）的header里必须有`cs-app-id`字段，登录后必须有`cs-token`字段
* 除密码外，其它所有字符串两端的空白字符都应去掉
* 服务器支持表单与JSON-BODY两种方式提交请求，服务器依据`Content-Type`识别请求的类型。
* 表单方式: `Content-Type:multipart/form-data`
* Json方式: `Content-Type:application/json;charset=UTF-8`
* 关于大小写：客户端上传的密码MD5应为小写；`cs-app-id`值区分大小写。
* 客户端须在请求中声明支持Gzip压缩，参考[Android的gzip处理](http://stackoverflow.com/questions/1573391/android-http-communication-should-use-accept-encoding-gzip)
* 关于日期与时间，分两种表示方式：
* 使用字符串表示时，**采用北京时区（GMT+8）**，日期格式为:`yyyy-MM-dd`, 时间格式为`yyyy-MM-dd HH:mm:ss`
* 也可用长整型unix timestamp表示时间。除非特别说明，timestamp均精确到毫秒


建议，
* 不要传值为空白的参数
* 客户端实现基本参数校验，如字符长度，格式检查
* 客户端应该有处理token过期的机制。(token可能因超时，或同一账号过多设备登录等失效）

## company_id的问题

注册与登录时必须提供`company_id`。且注意区分用于登录的`company_id`与获取称信息的`company_id`

## 标志类(flag)参数的用法
对于标志类参数，服务器依据此参数是否存在值判断true或false. 参考`/account/login`的`wdata`参数的用法

## header中的字段
> 服务器不区分字段名称的大小写。客户端解析时建议忽略**名称**的大小写。


| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| cs-app-id | 即原php服务器的app_id | 是 |
| cs-token |  登录或注册成功后返回的token  | 登录后必须 |
| cs-token-expirytime | token失效时间戳 | 否（只出现在登录、注册成功时服务器返回的header中） |
| cs-device-id | imei或等同唯一识别码 | 否 |
| user-agent | 客户端信息.   | 否 |
| cs-device | 手机型号及版本信息   | 否 |
| cs-app | app型号   | 否 |
| cs-scale | 秤的型号   | 否 |

**`user-agent, cs-device, cs-app, cs-scale`将用于后台统计设备信息**

`user-agent`格式为`package.name/version (platform; device model; platfrom-version)`如,
```
com.chipsea.btcontrol.en/1.4.6 (Android;SM-A300FU;5.0.2)
btWeigh/1.4.6 (iPhone; iOS 9.2.1; Scale/2.00)
```

# 账号相关


## 修改账号普通信息(签名、显示类别、称重与长度单位、App背景) POST /account

| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | --------: |
| signature | 社区的签名  | 否 |
| `mtypes` | `weight`, `bp`, `bsl`, `food`, `exercise`, `all`之一，或前者的组合（使用逗号拼接） | 否 |
| length_unit | 长度单位  | 否 |
| weight_unit | 称重单位  | 否 |
| appbg | App背景图名称。由后面的上传资源接口返回。  | 否 |

上述参数必须至少提供一个。

```bash
curl  -v \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: c61cf98a6dcf6bc4455f1db67301a6d62cfc251e43384c52ae15fb7cccb1f1cf ' \
-H 'Content-Type: application/json' \
-X POST \
-d '{
"signature" : "好好增肥",
"mtypes":"exercise,food",
"appbg": "000dc61ec56c09c6903453059080f62f"
}' \
http://[base-api]/account
```

返回

```json
{"code":200}
```

## 获取账号信息(签名、显示类别、称重与长度单位、App背景等等) GET /account

```bash
curl  -v \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: c61cf98a6dcf6bc4455f1db67301a6d62cfc251e43384c52ae15fb7cccb1f1cf ' \
http://[base-api]/account
```

返回除密码外当前账号的所有信息:

```json
{
"code": 200,
"data": {
"haier": "15019484424",
"phone": "15019484424",
"last_login": "2017-07-12",
"signature": "好好增肥",
"days": 0,
"weight_unit": 1400,
"company_id": 15,
"qq": "",
"sina_blog": "",
"grade_id": 0,
"register_time": "2017-03-20 17:41:13",
"id": 92043,
"length_unit": 1400,
"mtypes": "exercise,food"
}
}
```


## haier登录 POST /account/login

| 名称 | 描述 |
| ------------- | :------------- |
| haier | haier的账号（手机号或用户名等） |
| access_token | qq、新浪、微信、haier的`access_token` |
| company_id | 固定为15 |
| wdata | flag: 存在时返回account,role,remind数据，否则只返回account数据  |


```bash
curl  \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'Content-Type: application/json' \
-X POST \
-d '{"haier":"18856877767", "access_token":"7976511a-09a6-44b4-a608-abcdeffffff", "company_id": 15, "wdata": "t"}' \
https://hrznc.haier.net/account/login
```

返回的角色按role.id升序排列。role.id最小的角色即为主角色。

```json
{
"code": 200,
"data": {
"account": {
"company_id": 15,
"days": 0,
"grade_id": 0,
"haier": "18856877767",
"id": 208344,
"last_login": "2018-12-05",
"length_unit": 1400,
"mtypes": "weight",
"qq": "",
"register_time": "2018-12-05 08:31:52",
"signature": "",
"sina_blog": "",
"status": 1,
"weight_unit": 1400
},
"flash_remind": [
{
"account_id": 208344,
"fri_open": 0,
"id": 10120,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "8:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
},
{
"account_id": 208344,
"fri_open": 0,
"id": 10121,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "12:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
},
{
"account_id": 208344,
"fri_open": 0,
"id": 10122,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "17:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
}
],
"remind": [
{
"account_id": 208344,
"fri_open": 0,
"id": 139723,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "8:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
},
{
"account_id": 208344,
"fri_open": 0,
"id": 139724,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "12:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
},
{
"account_id": 208344,
"fri_open": 0,
"id": 139725,
"is_open": 0,
"mon_open": 0,
"once_open": 1,
"remind_time": "17:00",
"sat_open": 0,
"sun_open": 0,
"thu_open": 0,
"tue_open": 0,
"wed_open": 0
}
],
"role": [
{
"account_id": 208344,
"birthday": "1999-09-09",
"create_time": "2018-11-05 08:32:37",
"current_state": 1,
"height": 165,
"icon_image_path": "",
"id": 1995509,
"nickname": "大庆",
"role_type": 0,
"sex": "女",
"weight_goal": "53.0"
}
]
},
"msg": "OK"
}
```

登录成功后，返回的header中会包含如下的token信息:
```java
// token
Cs-Token:8f52ed8aef48ce8ffff65a1a3372c65c57898fa85f9d44b6ba14acab09c91b69
// 失效时间
Cs-Token-Expirytime:1439552877720
```



## 获取自己或他人的总积分 GET /jifen


| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | --------: |
| account_id | 账号id  | 否. 不提供时查询当前用户，提供时查询其它用户 |


```bash
curl -v \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
'http://[base-api]/jifen?account_id=158271'
```

返回如下

```json
{"code":200,"data":10}
```

## 获取积分记录 GET /jifen/list

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| account_id | 账号ID | 否，默认为当前用户 |
| cnt | 条目数 | 否，默认10 |
| lastid | 上次最小id | 否，默认最新一条记录id |


```bash
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
'http://[base-api]/jifen/list?account_id=158271&cnt=5&lastid=212223'
```


返回

```json
{
"code": 200,
"data": [
{
"id": 1,
"score": 10,
"source": 6,
"date": "2017-07-18",
"ts": 1500346980570,
"account_id": 158271
}
]
}
```

其中`score`为某次事件获得的分数，`source`为来源：


| source | 描述 | 得分 | 频次 |
| ------------- | :------------- | ----- | ----- |
| 101  | 首次注册登录成功 | 20 | 1 |
| 102  | 首次绑定体脂秤 | 20 | 1 |
| 103  | 首次绑定营养秤 | 20 | 1 |
| 104  | 签到 | 5 | 1次/天 |
| 105  | 打卡（无图） | 20 | 1次/周 |
| 106  | 打卡（有图） | 30 | 1次/周 |
| 107  | 评论打卡或文章 | 10 | 1次/周 |
| 108  | 上传视频并通过审核 | 30 | 1次/周 |
| 109  | 购买并确认收货 | 50 | 不限 |
| 110  | 评为达人 | 由管理员指定 | 不限 |
| 111  | 支付订单 | 支付订单时消耗的分值，为负数 | 不限 |
| 112  | 积分赠送 | 后台管理导入，分值由管理员指定 | 不限 |
| 113  | 退货退积分 | 退还原扣除的积分 | 不限 |

<!-- ``` -->
<!-- 1 称重 -->
<!-- 2 打卡 -->
<!-- 3 连打7天 -->
<!-- 4 连打21天 -->
<!-- 5 连打100天 -->
<!-- 6 打卡成达人 -->
<!-- 7 打卡参与活动 -->
<!-- 8 邀请注册 -->
<!-- ``` -->



## 申请积分兑换 PUT /jifen/coupon

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| coupon | 兑换码 | 是 |


```bash
curl  \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 3d6199a1c94d949e1736f0a8424323a177f9419a3b2b4c289635e525edf2bff5' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"coupon":"xxxxyyyyyyzzz"}' \
https://hrznc.haier.net/jifen/coupon
```

## 获取积分兑换结果列表 GET /jifen/coupon

```bash
curl  \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 3d6199a1c94d949e1736f0a8424323a177f9419a3b2b4c289635e525edf2bff5' \
-X GET \
http://localhost:3333/jifen/coupon
```

返回如下，其中

* `status`为`'accept (兑换成功), 'pending'（审核中）, 'reject'（失败）`。
* `reason`为文字描述成功或失败的原因
* `ts`申请时间

```json
{
"code": 200,
"data": [
{
"aid": 195966,
"coupon": "XXXXYYYYYYZZZ",
"id": 1949,
"reason": "",
"score": 0,
"status": "pending",
"ts": 1545615620204
}
],
"msg": "OK"
}
```



# 家庭成员

## 添加角色 PUT /role/:nickname

在当前账号下添加角色（家底成员）。一个账号最多只能包含7名成员。



注意：要上传头像时，须用表单方式提交请求。

| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | --------: |
| nickname | 昵称,在同一账号中必须唯一  | 是 |
| height | 身高(cm)  | 是 |
| birthday | 生日，格式`yyyy-MM-dd`  | 是 |
| sex | 性别: '男','女'  | 是 |
| icon | 头像  | 否  |
| weight_goal | 目标体重  | 否 |
| role_type | 整型角色类型，客户端定义  | 否，默认为0 |

```bash
curl -k -v \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'user-agent:com.chipsea.btcontrol.en/1.4.9 (Android;SAMSUNG-SM-G920A;6.0.1)' \
-H 'cs-token:3cc4b355f3deaa02da8da8d831795bcf6fea979c2b544c2396be3440664c2b68' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"birthday": "1999-07-09", "sex":"男", "height": 178 }' \
'http://[base-api]/role/%2Cafda'
```



调用成功后返回`data`为`role_id`，如下，
```json
{
"code": 200,
"data": 178
}
```

## 更新角色 POST /role/:roleid
修改当前登录用户下，此`roleid`的成员属性。



**注意：**
* 要上传头像时，须用表单方式提交请求。
* 只提交需要修改的字段。例如只需修改身高时，只提交`height`字段。


| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | --------: |
| nickname | 新昵称  | 否 |
| height | 身高(cm)  | 否 |
| birthday | 生日，格式`yyyy-MM-dd`  | 否 |
| sex | 性别: '男','女'  | 否 |
| icon | 头像文件 | 否  |
| weight_goal | 目标体重  | 否 |
| weight_init | 初始体重  | 否 |
| role_type | 整型角色类型，客户端定义  | 否 |


调用成功后返回结果如下，
```json
{"code":200}
```

## 删除角色 DELETE /role/:roleid
根据`role_id`删除成员。不允许删除此账号role.id最小的角色（即主角色）。



```bash
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token:6fabcddaswwdkwwkks19eb8f2dd5485e46eb7423ca2a88a61bd204f0f' \
-H 'Content-Type: application/json' \
-X DELETE \
http://[base-api]/role/273
```

调用成功后返回结果如下，
```json
{"code":200}
```

## 获取角色信息 GET /role/:roleid
获取成员信息。

> 用户头像请用本接口返回的`icon_image_path`，调用`/icon/:path`下载。


| 名称 | 描述 | 是否必须 |
| ------------- | :------------- | --------: |
| roleid | 角色ID | 是 |

调用成功后返回结果如下，
```json
{
"code": 200,
"data": {
"current_state": 1,
"sex": "男",
"weight_goal": 0,
"account_id": 99,
"icon_image_create_time": null,
"nickname": "fool",
"birthday": "1983-07-08T16:00:00Z",
"icon_image_path": "943b93c616dfe5bea181b8a5039f6372",
"sync_time": null,
"id": 173,
"create_time": "2015-09-11T02:14:16Z",
"modify_time": "2015-09-11T04:41:52Z",
"height": 188,
"period_time": null
}
}
```

## 获取当前账号下所有的角色信息 GET /roles

```bash
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 3cc4b355f3deaa02da8da8d831795bcf6fea979c2b544c2396be3440664c2b68 ' \
-H 'Content-Type: application/json' \
-H 'cs-device: Android;SM-A300FU;5.0.2' \
-H 'cs-scale: Hair;1.292' \
-H 'cs-app: com.haier/2.12.2' \
-X GET \
http://[base-api]/roles
```


返回示例：

```json
{
"code": 200,
"data": [
{
"current_state": 1,
"sex": "女",
"weight_goal": "0.0",
"account_id": 161705,
"nickname": "123456",
"birthday": "2000-06-15",
"icon_image_path": "",
"id": 276476,
"create_time": "2017-07-07 17:48:38",
"role_type": 0,
"height": 160
},
{
"current_state": 1,
"sex": "男",
"account_id": 161705,
"nickname": ",afda",
"birthday": "1999-07-09",
"icon_image_path": "",
"id": 276477,
"create_time": "2017-07-14 15:19:52",
"role_type": 0,
"height": 178
}
]
}
```




## 下载资源(头像、打卡图片、App背景) GET /icon/:path

`http://[base-api]/icon/654350ea6f443e5db4d9fabc77e96a70`

返回头像图片。文件不存在时返回HTTP状态404。

文件类型可以从Response header中获取，如下
```ini
Content-Type:image/jpeg
Date:Mon, 14 Sep 2015 02:53:19 GMT
Server:Jetty(7.6.13.v20130916)
```

# 测量数据与运动、饮食记录

## 添加数据 PUT /mdata

> 本操作不具有原子性。

在当前账号下添加一条或多条测量数据。

`mdata`数组

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | :--------: |
| `mdata` | 测量结果数组，测量结果的数据结构见下表 | 是 |


**体重测量结果字段**

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | :--------: |
| `mtype`  | 测量类型：`weight` | 否，但建议2.0后版本提供 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| `weight`  |  | 是 |
| `weight_time`  |  | 是 |
| axunge  | 身体脂肪率 | 否 |
| bmi  |  | 否 |
| body_age  | 正整数  | 否 |
| bone  |  | 否 |
| metabolism  | 基础代谢 | 否 |
| muscle  |  | 否 |
| viscera  | 内脏脂肪 | 否 |
| water  |  | 否 |
| r1  | 电阻值，浮点数 | 否 |
| age  | 年龄, tinyint | 否，秤数据 |
| sex  | 性别, tinyint，1或0 | 否，秤数据 |
| height  | 身高, tinyint | 否，秤数据 |
| rn8  | 电极电阻, varchar(200)，为逗号分开的浮点数 | 否，秤数据 |
| resistance  | V10电阻值，varchar(512)，电阻值字节数组经HEX编码 | 否，V10秤数据 |
| iseightr  | 是否8电极  | `y`是，其它值表示否 |



* 血压(bp)测量结果字段

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | :--------: |
| `mtype`  | 测量类型：`bp` | 是 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| `sys`  | 收缩压，integer | 是 |
| `dia`  | 舒张压，integer | 是 |
| hb  | 心率，integer | 否 |
| `measure_time`  |  | 是 |

* 血糖(bsl)测量结果字段

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | :--------: |
| `mtype`  | 测量类型：`bsl` | 是 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| `bsl`  | 血糖, float(5,2) | 是 |
| `description`  | ~~`pre-meal`:饭前，`post-meal`:饭后~~ `1`:空腹， `2`:早餐后，`3`:午饭前, `4`:午饭后, `5`:晚饭前,`6`:晚饭后 | ~~是~~ |
| `measure_time`  |  | 是 |


* 饮食(food)数组字段

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| `mtype`  | 测量类型：`food` | 是 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| name |  | 是 |
| quantity |  | 是 |
| food_id | 食材数据库的id  | 是 |
| date | 格式为`yyyy-MM-dd HH:mm:ss`的字符串 | 是 |
| unit | `g`, `ml`或`json`对象或`json`字符串  | 是 |
| calory | kCal  | 是 |
| ftype | `breakfast`,`lunch`,`dinner`,`snacks`,`unknown`.  | 是 |
| metabolism  | 基础代谢 | 是 |

* 运动（exercise）字段

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| `mtype`  | 测量类型：`exercise` | 是 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| name |  | 是 |
| duration | 分钟数 | 是 |
| date | 格式为`yyyy-MM-dd`的字符串 | 是 |
| ex_id |   | 是 |
| calory | kCal  | 是 |
| metabolism  | 基础代谢 | 是 |

* 课程训练（training）字段

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | ----- |
| `mtype`  | 测量类型：`training` | 是 |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| tid | 课程（tutorial）id | 是 |
| measure_time | 训练完成时间。格式为`yyyy-MM-dd HH:mm:ss`的字符串 | 是 |
| calory | 当前课程消耗的能量kCal  | 是 |
| metabolism  | 基础代谢 | 是 |

```bash
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 3cc4b355f3deaa02da8da8d831795bcf6fea979c2b544c2396be3440664c2b68 ' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"mdata": [{"axunge": 28, "viscera": 1, "metabolism": 473, "water": 99, "weight_time": "2015-07-27 09:25:30", "role_id": 276476, "body_age": 0, "bmi": 24, "weight": 1269.0, "muscle": 33, "bone": 3.4}, {"axunge": 28, "viscera": 1, "metabolism": 473, "water": 99, "weight_time": "2015-07-27 09:29:30", "role_id": 276477, "body_age": 0, "bmi": 24, "weight": 69, "muscle": 33, "bone": 3.4, "age": 88}]}' \
http://[base-api]/mdata

血压、血糖等
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 2dae556ecf6d207181ffd019d048ef01b6ed9e207e714937815fab290fa420fd  ' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"mdata": [{"role_id": 1209, "mtype": "bp", "sys": "120", "dia": "80", "hb": "80", "measure_time": "2015-07-27 09:22:22"},{"role_id": 1209, "mtype": "bsl", "bsl": 2.22, "description": "2", "measure_time": "2015-07-27 03:22:22"}, {"mtype": "weight","axunge": 28, "viscera": 1, "metabolism": 473, "water": 99, "weight_time": "2015-07-27 09:29:30", "role_id": 1209, "body_age": 0, "bmi": 24, "weight": 69, "muscle": 33, "bone": 3.4}]}' \
http://[base-api]/mdata

配餐/饮食
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 2323281aalk1136f0a8424323a177f9419a3b2b4c289635e525edf2bff5  ' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"mdata": [{"metabolism": 0,"mtype":"food","food_id":10203, "name":"雪花鱼丝羹", "quantity":100,"unit":"g","calory":32.5, "ftype":"unknown", "date":"2016-10-23 22:00:01", "role_id":354726}, {"metabolism": 0,"mtype":"food","food_id":1, "name":"小麦", "quantity":100,"unit":"g","calory":339.0, "ftype":"unknown", "date":"2016-10-23 22:00:01", "role_id":354726}]}' \
http://[base-api]/mdata



运动与饮食
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 2dae556ecf6d207181ffd019d048ef01b6ed9e207e714937815fab290fa420fd  ' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"mdata": [{"metabolism": 1228,"mtype":"food","food_id":10203, "name":"雪花鱼丝羹", "quantity":100,"unit":"g","calory":32.5, "ftype":"unknown", "date":"2016-10-23 22:00:01", "role_id":1209}, {"metabolism": 1228,"ex_id":1716, "name":"网球，双打", "duration":120,"calory":1292, "date":"2016-10-23", "role_id":1209, "mtype":"exercise"}]}' \
http://[base-api]/mdata

课程训练
curl -k \
-H 'cs-app-id:ebcad75de0d42a844d98a755644e30' \
-H 'cs-token: 2232322c94d949e1736f0a8424323a177f9419a3b2b4c289635e525edf2bff5  ' \
-H 'Content-Type: application/json' \
-X PUT \
-d '{"mdata": [{"mtype":"training", "role_id": 354726, "tid": 3, "measure_time": "2018-09-18 16:11:02", "calory": 1024, "metabolism": 1000}]}' \
http://[base-api]/mdata
```

返回如下，
```json
{
"data": [
{
"weight_time": "2015-07-27 09:25:30",
"id": 13437,
"role_id": 3
},
{
"weight_time": "2015-07-27 09:29:30",
"id": 13438,
"role_id": 3
}
],
"code": 200
}
```

**对于称重数据**，
* 相同称重时间的数据已经存在时：会返回已存在的数据，且`exists`值为`t`：
* 存在无效数据时：返回原数据，且`invalid`字段值为`t`

```json
{
"code": 200,
"data": [
{
"upload_time": "2016-09-19T01:25:39Z",
"account_id": 6385,
"mtype": "weight",
"bone": 3.4,
"muscle": 33,
"invalid": "t",
"weight": 1269.0,
"bmi": 24,
"body_age": 0,
"role_id": 10273,
"weight_time": "2015-07-27 09:25:30",
"water": 99,
"metabolism": 473,
"viscera": 1,
"axunge": 28
},
{
"exists": "t",
"upload_time": "2016-09-19 09:25:37",
"account_id": 6385,
"mtype": "weight",
"bone": "3.4",
"scaleproperty": 1,
"muscle": "33.0",
"weight": "69.0",
"id": 912642,
"bmi": "24.0",
"body_age": 0,
"productid": 0,
"role_id": 10273,
"weight_time": "2015-07-27 09:29:30",
"water": "99.0",
"metabolism": "473.0",
"viscera": "1.0",
"axunge": "28.0"
}
]
}
```

**另外两种数据类型(bp, bsl)**数据重复时，使用新上传数据覆盖服务器数据，返回的数据的`exists`值为`u`：。

```json
{
"data": [
{
"exists": "u",
"id": 39,
"role_id": 6189,
"measure_time": "2015-06-12 10:44:00"
}
],
"code": 200
}
```





## V10解析八电极数据 POST /weight/parse



参数，与普通测量字段相比，`sex age height resistance iseightr`是必须上传的：

| 参数名称 | 描述 | 是否必须 |
| ------------- | :------------- | :--------: |
| `role_id`  | 当前账号下的一使用者角色ID | 是 |
| `weight`  |  | 是 |
| `weight_time`  |  | 是 |
| axunge  | 身体脂肪率 | 否 |
| bmi  |  | 否 |
| body_age  | 正整数  | 否 |
| bone  |  | 否 |
| metabolism  | 基础代谢 | 否 |
| muscle  |  | 否 |
| viscera  | 内脏脂肪 | 否 |
| water  |  | 否 |
| r1  | 电阻值，浮点数 | 否 |
| age  | 年龄, tinyint | 是 |
| sex  | 性别, tinyint, 1或0 | 是 |
| height  | 身高, tinyint | 是 |
| resistance  | V10电阻值，varchar(512)，电阻值字节数组经HEX编码 | 否，V10秤数据 |
| iseightr  | 是否8电极. `y`是，其它值表示否  | 否 |


CHANGES.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [0.1.32](#0132)
- [0.1.22](#0122)
  - [升级说明](#%E5%8D%87%E7%BA%A7%E8%AF%B4%E6%98%8E)
  - [变更](#%E5%8F%98%E6%9B%B4)
    - [已知问题](#%E5%B7%B2%E7%9F%A5%E9%97%AE%E9%A2%98)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# 0.1.32
* 新增RPC接口
* 新增HTTP讲坛接口
* 生成随机用户
* 升级逻辑修改
* nginx/wordpress访问配置


# 0.1.22

## 升级说明
* 升级前需先执行合并国际版数据库脚本：


```sql
--- !!!执行前备份数据库!!!
source 006-merge-intl-okok.sql
call import_intl_data();
```

* 配置文件中添加参数`:sms-2`（普天短信）及`:pull-intl-okok-data-period`(定时导入PHP国际版数据)


## 变更
* 合并国际版数据库：导入用户及测量数据。
* company_id为2时，使用putian短信接口。
* `/latestapp`接口变更，见接口文档。

### 已知问题
* 使用PHP国际版接口更新、删除的数据不会并入okok数据库中。


src/chips/codes.clj
(ns chips.codes)


(def cds-map
  {:exception                    {:code 1 :msg "系统异常!"}
   :session-expired              {:code 2 :msg "未登录"}
   :auth-failed                  {:code 3 :msg "无效访问"}
   :params-conflict              {:code 4 :msg "参数有冲突！"}
   :invalid-access-token         {:code 5 :msg "Openid对应的access_token无效"}
   :api-not-implemented          {:code 6 :msg "此接口尚未开放"}
   :success                      {:code 200 :msg "OK"}
   :account_len_invalid          {:code 101 :msg "帐户名长度超过其允许的最大长度(注册)"}
   :account_no_exist             {:code 102 :msg "用户帐号不存在(登录)"}
   :register_account_exist       {:code 103 :msg "该帐户已经注册(注册) "}
   :register_failure             {:code 104 :msg "注册失败(注册)"}
   :account_password_error       {:code 105 :msg "密码错误(登录)"}
   :account_not_loggin           {:code 106 :msg "用户没有登录，服务器拒绝执行该请求(其他)"}
   ;; :email_send_failure           {:code 107 :msg   "邮件发送失败(用户反馈)"}
   :app_id_verify_error          {:code 108 :msg "app_id不正确"}
   :role_add_error               {:code 109 :msg "角色添加失败"}
   :role_already_exist           {:code 110 :msg "角色已经存在"}
   :role-not-exist               {:code 111 :msg "原角色不存在"}
   :role-already-deleted         {:code 112 :msg "角色已经删除"}
   :weigher_already_exist        {:code 113 :msg "设备已经存在"}
   :weigher_already_error        {:code 114 :msg "设备添加失败"}
   :weigher_del_error            {:code 115 :msg "添加删除失败"}
   :weigher_update_error         {:code 116 :msg "修改设备失败"}
   :role_find_error              {:code 117 :msg "用户查询失败"}
   :software_version_newest      {:code 118 :msg "软件已经是最新版本"}
   :role_feedback_err            {:code 119 :msg "暂时没有数据"}
   :image_create_err             {:code 120 :msg "图片生成失败"}
   :image_db_err                 {:code 121 :msg "数据库操作失败"}
   :role_feedback_invalid        {:code 122 :msg "用户反馈意见长度超过其允许的最大长度(用户反馈)"}
   :main-role-delete-not-allowed {:code 123 :msg "不能删除主账号"}
   :unbind-not-allowed           {:code 131 :msg "不能解绑主账号"}
   :already-bind                 {:code 132 :msg "已经绑定"}
   :vericode-invalid             {:code 133 :msg "验证码错误"}
   :verifycode_send_error        {:code 134 :msg "验证码发送失败"}
   :too-many-roles               {:code 135 :msg "家庭成员最多不能超过8人"}
   :required-param-not-exist     {:code 301 :msg "缺少必要参数"}
   :invalid-uid                  {:code 302 :msg "uid格式不合法"}
   :invalid-provider             {:code 303 :msg "登录类型不明，必须是sina_blog/qq/uid之一"}
   :invalid-phone                {:code 304 :msg "手机号格式不合法"}
   :invalid-param                {:code 305 :msg "参数不合法"}
   :request-interval-too-short   {:code 306 :msg "请求太频繁!"}
   :id-taken                     {:code 307 :msg "当前账号已被使用"}
   :pwds-equal                   {:code 308 :msg "新旧密码相同"}
   :invalid-sex                  {:code 309 :msg "性别不合法：男/女"}
   :unknown-action               {:code 310 :msg "未知操作"}
   :invalid-reminder-time        {:code 311 :msg "提醒时间无效"}
   :invalid-length-unit          {:code 312 :msg "长度单位无效"}
   :invalid-weight-unit          {:code 313 :msg "质量单位无效"}
   :invalid-company-id           {:code 314 :msg "companyid无效"}
   
   :openid-taken                    {:code 315 :msg "当前QQ或新浪账号已经在使用"}
   :invalid-email                   {:code 316 :msg "email格式不合法"}
   :invalid-mtype                   {:code 400 :msg "测量类别无效"}
   :emtpy-comment-content           {:code 401 :msg "评论内容为空"}
   :invalid-ptype                   {:code 402 :msg "统计时段无效"}
   :account-disabled                {:code 403 :msg "账号已被禁用"}
   :account-post-mblog-disabled     {:code 404 :msg "账号已被禁言"}
   :order-not-exists                {:code 405 :msg "订单不存在"}
   :order-invalid-state             {:code 406 :msg "订单正在处理中，不允许修改"}
   :product-removed                 {:code 407 :msg "商品已下架"}
   :invalid-payment-type            {:code 408 :msg "支付方式无效"}
   :not-enough-score                {:code 409 :msg "积分不足"}
   :order-status-invalid            {:code 410 :msg "订单状态无效"}
   :order-no-need-to-pay            {:code 411 :msg "订单无须支付"}
   :order-from-same-seller-required {:code 412 :msg "不同店铺的商品请分别下单！"}
   :coupon-exists                   {:code 413 :msg "兑换码无效或已使用！"}
   :product-not-exists              {:code 414 :msg "商品不存在！"}
   :not-enough-prod-to-return       {:code 415 :msg "退换商品数量超过实际购买数量！"}
   :invalid-order-status-to-return  {:code 416 :msg "订单完成后方可退换商品！"}
   :order-same-prod-should-group    {:code 417 :msg "同一订单相同商品不能分开！"}
   :invalid-order-status            {:code 418 :msg "当前订单状态下不允许此操作！"}
   })


(def ^:prviate err (RuntimeException. "Code name not exists!"))

(defn cds
  "Returns `(:code-name cds)` or throws an exception if the `:code-name`
  does not exist in `cds`"
  ([code-name]
   (if-let [c (get cds-map code-name)]
     c
     (throw err)))
  ([code-name alt-map]
   (merge (cds code-name) alt-map)))

(comment
  (run! (fn [{:keys [code msg]}]
          (println (format "| %d | %s |" code msg)))
        (sort-by :code (vals cds-map))))
